l10n_it_edi_doi_extension/models/purchase_order.py

```diff
@@ -0,0 +1,310 @@
+            if not declaration:
+                return
+            partner = order.partner_id.commercial_partner_id
+            errors = declaration._get_validity_warnings(
+                order.company_id,
+                partner,
+                order.currency_id,
+                order.l10n_it_edi_doi_date,
+                only_blocking=True,
+                sales_order=True,
+            )
+            if errors:
+                raise ValidationError("\n".join(errors))
+
+    def action_open_declaration_of_intent(self):
+        self.ensure_one()
+        return {
+            "name": _("Declaration of Intent for %s", self.display_name),
+            "type": "ir.actions.act_window",
+            "view_mode": "form",
+            "res_model": "l10n_it_edi_doi.declaration_of_intent",
+            "res_id": self.l10n_it_edi_doi_id.id,
+        }
+
+    def _l10n_it_edi_doi_get_amount_not_yet_invoiced(
+        self, declaration, additional_invoiced_qty=None
+    ):
+        """
+        Consider purchase orders in self that use declaration of intent `declaration`.
+        For each purchase order we compute the amount that is tax exempt due to the
+        declaration of intent (line has special declaration of intent tax applied)
+        but not yet invoiced. For each line of the PO we i.e. use the not yet invoiced
+        quantity to compute this amount. The aforementioned quantity is computed from
+        field `qty_invoiced` and parameter `additional_invoiced_qty`
+        Return the sum of all these amounts on the SOs.
+        :param declaration:             We only consider purchase orders using
+                                        Declaration of Intent `declaration`.
+        :param additional_invoiced_qty: Dictionary (purchase order line id -> float)
+                                        The float represents additional invoiced amount
+                                        qty for the purchase order. This can be used
+                                        to simulate posting an already linked invoice.
+        """
+        if not declaration:
+            return 0
+
+        if additional_invoiced_qty is None:
+            additional_invoiced_qty = {}
+
+        tax = declaration.company_id.l10n_it_edi_doi_bill_tax_id
+        if not tax:
+            return 0
+
+        not_yet_invoiced = 0
+        for order in self:
+            if declaration != order.l10n_it_edi_doi_id:
+                continue
+
+            order_lines = order.order_line.filtered(
+                lambda line: line.taxes_id.ids == tax.ids
+            )
+            order_not_yet_invoiced = 0
+            for line in order_lines:
+                price_reduce = line.price_unit * (1 - (line.discount or 0.0) / 100.0)
+                qty_invoiced = line.qty_invoiced
+                if line.ids and additional_invoiced_qty:
+                    qty_invoiced += additional_invoiced_qty.get(line.ids[0], 0)
```

Comment:
Ho visto che nel metodo analogo di `l10n_it_edi_doi` è come hai scritto, ma non capisco che senso abbia quindi ti propongo questa
```suggestion
                if line.id in additional_invoiced_qty:
                    qty_invoiced += additional_invoiced_qty.get(line.id, 0)
```
In ogni caso, mi pare che questo metodo non venga mai chiamato on `additional_invoiced_qty`.

Reply:
:+1: